xmtp_mls/src/storage/encrypted_store/db_connection.rs

```diff
@@ -19,37 +26,81 @@ pub type DbConnection = DbConnectionPrivate<sqlite_web::connection::WasmSqliteCo
 // callers should be able to accomplish everything with one conn/reference.
 #[doc(hidden)]
 pub struct DbConnectionPrivate<C> {
-    inner: Arc<Mutex<C>>,
+    // Connection with read-only privileges
+    read: Option<Arc<Mutex<C>>>,
+    // Connection with write privileges
+    write: Arc<Mutex<C>>,
+    // Is any connection (possibly this one) currently in a transaction?
+    transaction_lock: Arc<Mutex<()>>,
+    // Is this particular connection in a transaction?
+    in_transaction: Arc<AtomicBool>,
 }
 
 /// Owned DBConnection Methods
 impl<C> DbConnectionPrivate<C> {
     /// Create a new [`DbConnectionPrivate`] from an existing Arc<Mutex<C>>
-    pub(super) fn from_arc_mutex(conn: Arc<Mutex<C>>) -> Self {
-        Self { inner: conn }
+    pub(super) fn from_arc_mutex(
+        write: Arc<Mutex<C>>,
+        read: Option<Arc<Mutex<C>>>,
+        transaction_lock: Arc<Mutex<()>>,
+    ) -> Self {
+        Self {
+            read,
+            write,
+            transaction_lock,
+            in_transaction: Arc::new(AtomicBool::new(false)),
+        }
     }
 }
 
 impl<C> DbConnectionPrivate<C>
 where
     C: diesel::Connection,
 {
+    pub(crate) fn start_transaction<Db: XmtpDb<Connection = C>>(
+        &self,
+    ) -> Result<TransactionGuard<'_>, StorageError> {
+        let guard = self.transaction_lock.lock();
+        let mut write = self.write.lock();
+        <Db as XmtpDb>::TransactionManager::begin_transaction(&mut *write)?;
+        self.in_transaction.store(true, Ordering::SeqCst);
+
+        Ok(TransactionGuard {
+            _mutex_guard: guard,
+            in_transaction: self.in_transaction.clone(),
+        })
+    }
+
     /// Do a scoped query with a mutable [`diesel::Connection`]
     /// reference
-    pub(crate) fn raw_query<T, E, F>(&self, fun: F) -> Result<T, E>
+    pub(crate) fn raw_query_read<T, E, F>(&self, fun: F) -> Result<T, E>
     where
         F: FnOnce(&mut C) -> Result<T, E>,
     {
-        let mut lock = self.inner.lock();
+        let mut lock = if self.in_transaction.load(Ordering::SeqCst) {
+            self.write.lock()
+        } else if let Some(read) = &self.read {
+            read.lock()
+        } else {
+            self.write.lock()
+        };
+
         fun(&mut lock)
     }
 
-    /// Internal-only API to get the underlying `diesel::Connection` reference
-    /// without a scope
-    /// Must be used with care. holding this reference while calling `raw_query`
-    /// will cause a deadlock.
-    pub(super) fn inner_mut_ref(&self) -> parking_lot::MutexGuard<'_, C> {
-        self.inner.lock()
+    /// Do a scoped query with a mutable [`diesel::Connection`]
+    /// reference
+    pub(crate) fn raw_query_write<T, E, F>(&self, fun: F) -> Result<T, E>
+    where
+        F: FnOnce(&mut C) -> Result<T, E>,
+    {
+        let _guard;
+        // If this connection is not in a transaction
+        if !self.in_transaction.load(Ordering::SeqCst) {
```

Comment:
what happens if we _are_ in a transaction, and a worker/thread not in a transaction tries to write with the connection that started a transaction

Reply:
This thread should own the guard belonging to the lock on 101, and will have to wait until we're done.